stonesoup/hypothesiser/gaussianmixture.py

```diff
@@ -0,0 +1,116 @@
+            # Get hypotheses for that component for all measurements
+            component_hypotheses = self.hypothesiser.hypothesise(component,
+                                                                 detections,
+                                                                 timestamp)
+            for hypothesis in component_hypotheses:
+                if isinstance(component, TaggedWeightedGaussianState):
+                    hypothesis.prediction = \
+                        TaggedWeightedGaussianStatePrediction(
+                            tag=component.tag if component.tag != "birth"
+                            else None,
+                            weight=component.weight,
+                            state_vector=hypothesis.prediction.state_vector,
+                            covar=hypothesis.prediction.covar,
+                            timestamp=hypothesis.prediction.timestamp
+                            )
+                else:
+                    hypothesis.prediction = WeightedGaussianStatePrediction(
+                        weight=component.weight,
+                        state_vector=hypothesis.prediction.state_vector,
+                        covar=hypothesis.prediction.covar,
+                        timestamp=hypothesis.prediction.timestamp
+                    )
+            # Create Multiple Hypothesis and add to list
+            if len(component_hypotheses) > 0:
+                hypotheses.append(MultipleHypothesis(component_hypotheses))
+
+        # Reorder list of MultipleHypothesis so that they are ordered
+        # by detection, not component
+        if self.order_by_detection:
+            single_hypothesis_list = list()
+            # Retrieve all single hypotheses
+            for multiple_hypothesis in hypotheses:
+                for single_hypothesis in multiple_hypothesis:
+                    single_hypothesis_list.append(single_hypothesis)
+            reordered_hypotheses = list()
+            # Get miss detected components
+            miss_detections_hypothesis = MultipleHypothesis(
+                [x for x in single_hypothesis_list if not x])
+            for detection in detections:
+                # Create multiple hypothesis per detection
+                indices \
+                    = [x for x in range(len(single_hypothesis_list))
+                        if single_hypothesis_list[x].measurement == detection]
+                detection_multiple_hypothesis = \
+                    MultipleHypothesis(list(
+                                map(single_hypothesis_list.__getitem__,
+                                    indices)))
+                # Add to new list
+                reordered_hypotheses.append(detection_multiple_hypothesis)
+            # Add miss detected hypothesis to end
+            reordered_hypotheses.append(miss_detections_hypothesis)
+            # Assign reordered list to original list
+            hypotheses = reordered_hypotheses
```

Comment:
Looks like you can get [hypothesis by detection using `__getitem__` interface](https://github.com/dstl/Stone-Soup/blob/4715da2f22f694f384953d9db66499d1b6bf75c7/stonesoup/types/multihypothesis.py#L83). This may allow you to simplify the below.

Tests are still passing with this:
```python
        if self.order_by_detection:
            reordered_hypotheses = []
            # Create a multiple hypothesis per detection
            for detection in detections:
                reordered_hypotheses.append(MultipleHypothesis(
                    [multi_hypothesis[detection]
                     for multi_hypothesis in hypotheses
                     if detection in multi_hypothesis]))
            # Add miss detected hypothesis to end
            reordered_hypotheses.append(MultipleHypothesis(
                [hypothesis
                 for multi_hypothesis in hypotheses
                 for hypothesis in multi_hypothesis
                 if not hypothesis]))
            # Assign reordered list to original list
            hypotheses = reordered_hypotheses
```

Reply:
OK, I think I've made it a bit clearer in my most recent commit using list comprehension.